### PR TITLE
Reply feature for Doctor notes 

### PR DESCRIPTION
**Pull request by UdaySagar-Git**
### Associated Issue

Frontend PR: https://github.com/coronasafe/care_fe/pull/7805

### Architecture changes

- Remove this section if not used

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

File: care/facility/api/serializers/patient.py [hunk 1]
```diff
@@ -539,6 +546,11 @@ def create(self, validated_data):
             # If the user is not a doctor then the user type is the same as the user type
             validated_data["user_type"] = user_type
 
+        if validated_data.get("reply_to"):
+            validated_data["reply_to"] = PatientNotes.objects.get(
+                external_id=validated_data["reply_to"]
+            )
+
```
Comment (shivankacker) [inline care/facility/api/serializers/patient.py]: Add conditions in `get` to check if the `reply_to` belongs to the same thread and consultation

**Review by shivankacker — CHANGES_REQUESTED** [review]

File: care/facility/api/serializers/patient.py (hunk 1 above)
Comment (UdaySagar-Git) [reply care/facility/api/serializers/patient.py]: added

**Review by shivankacker — APPROVED** [review]

**Comment by nihal467** [comment]
@UdaySagar-Git fix the lint and test in the PR

**Comment by UdaySagar-Git** [comment]
@nihal467 fixed !

File: care/facility/api/serializers/patient.py [hunk 2]
```diff
@@ -518,6 +518,13 @@ class PatientNotesSerializer(serializers.ModelSerializer):
     thread = serializers.ChoiceField(
         choices=PatientNoteThreadChoices, required=False, allow_null=False
     )
+    reply_to = serializers.UUIDField(required=False, allow_null=True)
+
+    def to_representation(self, instance):
+        ret = super().to_representation(instance)
+        if ret.get("reply_to"):
+            ret["reply_to"] = instance.reply_to.external_id
+        return ret
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]: use externalidrelatedfield

**Review by sainak — CHANGES_REQUESTED** [review]

**Comment by sainak** [comment]
@UdaySagar-Git how will you ensure that the notes that you fetch that are replies include the parent in response?, also what if the notes have replies, but it were not fetched in the current response

cc @rithviknishad

**Comment by UdaySagar-Git** [comment]
@sainak , thanks for pointing that out. It won't fetch the parent if there are more notes. Instead of sending just the ID, sending the complete parent object along with the note would solve this. Should I consider that and make changes in frontend according to it?

**Comment by rithviknishad** [comment]
@UdaySagar-Git we could add a `ReplyToPatientNoteSerializer` containing a lesser number of fields, just the required ones.

File: care/facility/api/serializers/patient.py [hunk 3]
```diff
@@ -518,6 +532,17 @@ class PatientNotesSerializer(serializers.ModelSerializer):
...
+        queryset=PatientNotes.objects.all(),
+        required=False,
+        allow_null=True,
+    )
+
+    def to_representation(self, instance):
+        ret = super().to_representation(instance)
+        if instance.reply_to:
+            ret["reply_to"] = ReplyToPatientNoteSerializer(instance.reply_to).data
+        return ret
```
Comment (rithviknishad) [inline care/facility/api/serializers/patient.py]: We could simply do this instead right?


```suggestion
    reply_to_object = ReplyToPatientNoteSerializer(source="reply_to", read_only=True)
```

**Review by rithviknishad — CHANGES_REQUESTED** [review]
Other than this, LGTM. 

Tests would be appreciated 😄

**Review by rithviknishad — APPROVED** [review]
Tests needed

**Comment by UdaySagar-Git** [comment]
@rithviknishad forgot to commit 😅 , 
just added

**Review by rithviknishad — APPROVED** [review]

File: care/facility/api/serializers/patient.py [hunk 4]
```diff
@@ -503,6 +503,20 @@ class Meta:
         exclude = ("patient_note",)
 
 
+class ReplyToPatientNoteSerializer(serializers.ModelSerializer):
+    created_by_object = UserBaseMinimumSerializer(source="created_by", read_only=True)
```
Comment (sainak) [inline care/facility/api/serializers/patient.py:505]:
```suggestion
    id = serializers.CharField(source="external_id", read_only=True)
    created_by_object = UserBaseMinimumSerializer(source="created_by", read_only=True)
```

the `id` in field would return the db `id`, we should avoid exposing it

**Review by sainak — CHANGES_REQUESTED** [review]
add `select_related("reply_to")` in the notes viewset

**Comment by nihal467** [comment]
LGTM

File: care/facility/api/viewsets/patient.py [hunk 5]
```diff
@@ -910,7 +910,7 @@ class PatientNotesViewSet(
 ):
     queryset = (
         PatientNotes.objects.all()
-        .select_related("facility", "patient", "created_by")
+        .select_related("facility", "patient", "created_by", "reply_to")
```
Comment (sainak) [inline care/facility/api/viewsets/patient.py]:
```suggestion
        .select_related("facility", "patient", "created_by", "reply_to", "reply_to__created_by")
```

**Review by sainak — APPROVED** [review]

File: care/facility/api/serializers/patient.py [hunk 6]
```diff
@@ -539,6 +561,17 @@ def create(self, validated_data):
             # If the user is not a doctor then the user type is the same as the user type
             validated_data["user_type"] = user_type
 
+        if validated_data.get("reply_to"):
+            reply_to_note = validated_data["reply_to"]
+            if reply_to_note.thread != validated_data["thread"]:
+                raise serializers.ValidationError(
+                    "Reply to note should be in the same thread"
+                )
+            if reply_to_note.consultation != validated_data.get("consultation"):
```
Comment (vigneshhari) [inline care/facility/api/serializers/patient.py:571]: consultation is not a required field, need better validations

**Comment by nihal467** [comment]
@UdaySagar-Git can you make the changes requested

File: care/facility/api/serializers/patient.py (hunk 6 above)
Comment (UdaySagar-Git) [reply care/facility/api/serializers/patient.py:571]: Could you please suggest what additional validations can be implemented?

**Comment by nihal467** [comment]
@UdaySagar-Git what is the update on the PR

**Comment by UdaySagar-Git** [comment]
@nihal467 this pr is ready for a final review

**Comment by UdaySagar-Git** [comment]
@vigneshhari can you please review this , added a extra validation check for consultation
since the doctor notes will be specfic to consultation and is required

**Review by rithviknishad — APPROVED** [review]

**Comment by UdaySagar-Git** [comment]
@nihal467 can you try re-running the tests now ?

**Review by khavinshankar — APPROVED** [review]

**Comment by vigneshhari** [comment]
@sainak re-review required

**Review by sainak — APPROVED** [review]

**Comment by bodhish** [comment]
@sainak why are we not merging this?

**Comment by sainak** [comment]
@bodhish this needs a final review from vignesh

**Pull request merged**